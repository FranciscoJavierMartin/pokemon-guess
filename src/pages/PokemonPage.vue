<template>
  <h1 v-if="loading">Loading</h1>
  <div v-else>
    <h1>Which Pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :options="pokemons" @selection="checkAnswer" />
    <div v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">New Game</button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import PokemonPicture from '@/components/PokemonPicture.vue';
import PokemonOptions from '@/components/PokemonOptions.vue';
import getPokemonsOptions from '@/helpers/getPokemonsOptions';
import { Pokemon } from '@/interfaces/pokemon';

export default defineComponent({
  components: { PokemonPicture, PokemonOptions },
  data() {
    return {
      pokemons: [] as Pokemon[],
      pokemon: {} as Pokemon,
      loading: true,
      showPokemon: false,
      showAnswer: false,
      message: '',
    };
  },
  methods: {
    async getPokemons(): Promise<void> {
      this.pokemons = await getPokemonsOptions();

      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemons[rndInt];

      this.loading = false;
    },
    checkAnswer(selectedId: number): void {
      this.showPokemon = true;
      this.showAnswer = true;
      if (selectedId === this.pokemon.id) {
        this.message = `Correct, ${this.pokemon.name}`;
      } else {
        this.message = `Wrong, it is ${this.pokemon.name}`;
      }
    },
    newGame(): void {
      this.showAnswer = false;
      this.showPokemon = false;
      this.loading = true;
      this.pokemons = [];
      this.pokemon = {} as Pokemon;
      this.message = '';
      this.getPokemons();
    },
  },
  mounted() {
    this.getPokemons();
  },
});
</script>
