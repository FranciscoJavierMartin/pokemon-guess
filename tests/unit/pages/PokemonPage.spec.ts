import { shallowMount, mount } from '@vue/test-utils';
import PokemonPage from '@/pages/PokemonPage.vue';
import { pokemons } from '../mocks/pokemons.mock';

describe('PokemonPage', () => {
  test('Snapshot', () => {
    const wrapper = shallowMount(PokemonPage);
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('mounted should be called', () => {
    const getPokemonsSpy = jest.spyOn(
      PokemonPage.methods as any,
      'getPokemons'
    );
    const wrapper = shallowMount(PokemonPage);
    expect(getPokemonsSpy).toHaveBeenCalled();
  });

  test('Snapshot when data is loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
          loading: false,
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should show PokemonPicture and PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          loading: false,
          message: '',
        };
      },
    });

    expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy();
    expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy();
    expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe(
      '25'
    );
    expect(
      wrapper.find('pokemon-options-stub').attributes('options')
    ).toBeTruthy();
  });

  test('tests with checkAnswer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          loading: false,
          message: '',
        };
      },
    });

    await wrapper.vm.checkAnswer(25);

    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find('h2').text()).toBe(`Correct, ${pokemons[0].name}`);
  });
});
