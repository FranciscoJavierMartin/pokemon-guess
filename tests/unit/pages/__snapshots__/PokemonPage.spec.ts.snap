// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Snapshot 1`] = `<h1>Loading</h1>`;

exports[`PokemonPage Snapshot when data is loaded 1`] = `
<div>
  <h1>Which Pokemon?</h1>
  <pokemon-picture-stub pokemonid="25" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub options="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
